Make the vendor random number generators optional

fwyzard opened the pull request:
If `alpaka_DISABLE_VENDOR_RNG` is defined, alpaka will not include support for the vendor-specific random number generators (NVIDIA cuRAN, AMD hipRAND/rocRAND, Indel DPL).

Instead, all accelerators use the default configuration of the Philox RNG.

fwyzard committed:
Make the vendor-specific random number generators optional

If ALPAKA_DISABLE_VENDOR_RNG is defined, disable the vendor-specific random
number generators (NVIDIA cuRAND, AMD rocRAND, Intel DPL), and use the
alpaka native implementation based on the Philox randon number generator.

fwyzard commented:
Now... how do I make it optional with a CMake flag ?

AuroraPerego commented:
> Now... how do I make it optional with a CMake flag ?

Something like this in `cmake/alpakaCommon.cmake` should work:
```cmake
option(alpaka_DISABLE_RANDOM "Disable the the random number generators" OFF)

# Compiler settings.
if(alpaka_DISABLE_RANDOM)
    target_compile_definitions(alpaka INTERFACE "ALPAKA_DISABLE_RANDOM")
endif()
```
and then the option `-Dalpaka_DISABLE_RANDOM` can be passed to `cmake`.

AuroraPerego commented:
Maybe also `randTest` should be enclosed in a `#ifndef ALPAKA_DISABLE_RANDOM` or skipped directly when the `cmake` flag is used adding to `test/unit/rand/CMakeLists.txt` something like:
```cmake
if (alpaka_DISABLE_RANDOM STREQUAL "ON")
    message(STATUS "randTest requires the random number generators libraries. Test disabled.")
    return()
endif ()
```

fwyzard committed:
Add the alpaka_DISABLE_VENDOR_RNG option

fwyzard committed:
Add missing include

fwyzard commented:
Looks like including Philox in the CUDA and HIP accelerators leads to a recursive loop of header files inclusion... :-(

fwyzard committed:
Forward declare AccGpuUniformCudaHipRt to avoid a dependency loop

fwyzard committed:
Add CUDA/HIP headers

Due to the forward declaration in PhiloxBaseTraits, CudaVectorArrayWrapper may
be included before the CUDA or HIP API files.

fwyzard committed:
Remove unused variable

j-stephan commented on include/alpaka/acc/AccGpuUniformCudaHipRt.hpp:
```diff
@@ -64,7 +65,12 @@ namespace alpaka
+#    ifdef ALPAKA_DISABLE_VENDOR_RNG
+        public rand::RandDefault,
+#    else
+          public rand::RandUniformCudaHipRand<TApi>
+        ,
```
This looks odd. Did `clang-format` do this?

fwyzard replied:
Yes... 
I changed the other files and moved the comma in front to avoid the ugly code, but I missed this one.

fwyzard replied:
Actually, maybe I did not push that change? I'll check.

fwyzard replied:
I missed a couple files, hopefully they look better now.

bernhardmgruber commented:
This PR may break the amalgamation, but I can fix this for you in a follow-up commit.

fwyzard commented:
> This PR may break the amalgamation, but I can fix this for you in a follow-up commit.

Thanks!
